keras_hub/src/models/qwen_moe/qwen_moe_decoder.py

```diff
@@ -0,0 +1,500 @@
+        layer_norm_epsilon=1e-5,
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        self.hidden_dim = hidden_dim
+        self.moe_intermediate_dim = moe_intermediate_dim
+        self.shared_expert_intermediate_dim = shared_expert_intermediate_dim
+        self.num_experts = num_experts
+        self.top_k = top_k
+        self.norm_topk_prob = norm_topk_prob
+        self.kernel_initializer = kernel_initializer
+        self.layer_norm_epsilon = layer_norm_epsilon
+
+    def build(self, decoder_sequence_shape):
+        self._sparse_feedforward_gate_dense = keras.layers.Dense(
+            self.num_experts,
+            kernel_initializer=clone_initializer(self.kernel_initializer),
+            use_bias=False,
+            dtype=self.dtype_policy,
+            name="sparse_feedforward_gate_dense",
+        )
+        self._sparse_feedforward_gate_dense.build(decoder_sequence_shape)
+
+        self.experts = [
+            QwenMoeMLP(
+                intermediate_dim=self.moe_intermediate_dim,
+                hidden_dim=self.hidden_dim,
+                kernel_initializer=self.kernel_initializer,
+                layer_norm_epsilon=self.layer_norm_epsilon,
+            )
+            for _ in range(self.num_experts)
+        ]
+        for expert in self.experts:
+            expert.build(decoder_sequence_shape)
+
+        self.shared_expert_dense = QwenMoeMLP(
+            intermediate_dim=self.shared_expert_intermediate_dim,
+            hidden_dim=self.hidden_dim,
+            kernel_initializer=self.kernel_initializer,
+            layer_norm_epsilon=self.layer_norm_epsilon,
+        )
+        self.shared_expert_dense.build(decoder_sequence_shape)
+
+        self.shared_expert_gate_dense = keras.layers.Dense(1, use_bias=False)
+        self.shared_expert_gate_dense.build(decoder_sequence_shape)
+        self.built = True
+
+    def call(self, hidden_states):
+        batch_size, seq_len, hidden_dim = hidden_states.shape
+        hidden_states = hidden_states.reshape(-1, hidden_dim)
+
+        router_logits = self._sparse_feedforward_gate_dense(hidden_states)
+
+        routing_weights = ops.softmax(router_logits, axis=1)
+        routing_weights, selected_experts = ops.top_k(
+            routing_weights, self.top_k
+        )
+
+        if self.norm_topk_prob:
+            routing_weights /= ops.sum(routing_weights, axis=-1, keepdims=True)
+
+        routing_weights = ops.cast(routing_weights, hidden_states.dtype)
+
+        final_hidden_states = ops.zeros(
+            (batch_size * seq_len, hidden_dim), dtype=hidden_states.dtype
+        )
+
+        expert_mask = ops.one_hot(
+            selected_experts, num_classes=self.num_experts
+        )
+        expert_mask = ops.transpose(expert_mask, axes=[2, 1, 0])
+
+        for expert_idx in range(self.num_experts):
+            expert_layer = self.experts[expert_idx]
+            idx, top_x = ops.where(expert_mask[expert_idx])
```

Comment:
will the ops.where return different output shape here on different forward passes? - if so that would not work with JAX XLA

Reply:
I don't think that this should be the case. Why do you think so?

Reply:
Whether the output shape is differing or not doesn't matter. `ops.where` won't work on JAX at all, if you don't provide x and y.

`ops.where` calls [`jnp.where`](https://docs.jax.dev/en/latest/_autosummary/jax.numpy.where.html). Check the note [here](https://docs.jax.dev/en/latest/_autosummary/jax.numpy.nonzero.html#jax.numpy.nonzero):

```
Because the size of the output of nonzero is data-dependent, the function is not compatible with JIT and other transformations. The JAX version adds the optional size argument which must be specified statically for jnp.nonzero to be used within JAX’s transformations.
```